server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/UndesiredAllocationsTracker.java

```diff
@@ -0,0 +1,179 @@
+    private volatile int maxUndesiredAllocationsToTrack;
+
+    UndesiredAllocationsTracker(ClusterSettings clusterSettings, TimeProvider timeProvider) {
+        this.timeProvider = timeProvider;
+        this.undesiredAllocationDurationLogInterval = new FrequencyCappedAction(timeProvider::relativeTimeInMillis, TimeValue.ZERO);
+        clusterSettings.initializeAndWatch(
+            DesiredBalanceReconciler.UNDESIRED_ALLOCATIONS_LOG_INTERVAL_SETTING,
+            undesiredAllocationDurationLogInterval::setMinInterval
+        );
+        clusterSettings.initializeAndWatch(
+            UNDESIRED_ALLOCATION_DURATION_LOG_THRESHOLD_SETTING,
+            value -> undesiredAllocationDurationLoggingThreshold = value
+        );
+        clusterSettings.initializeAndWatch(MAX_UNDESIRED_ALLOCATIONS_TO_TRACK, value -> this.maxUndesiredAllocationsToTrack = value);
+    }
+
+    /**
+     * Track an allocation as being undesired
+     */
+    public void trackUndesiredAllocation(ShardRouting shardRouting) {
+        assert shardRouting.unassigned() == false : "Shouldn't record unassigned shards as undesired allocations";
+        if (undesiredAllocations.containsKey(shardRouting) == false && undesiredAllocations.size() < maxUndesiredAllocationsToTrack) {
+            undesiredAllocations.put(shardRouting, timeProvider.relativeTimeInMillis());
+        }
+    }
+
+    /**
+     * Clear any tracking of the provided allocation
+     */
+    public void clear(ShardRouting shardRouting) {
+        undesiredAllocations.remove(shardRouting);
+    }
+
+    /**
+     * Clear any {@link ShardRouting} that are no longer present in the routing nodes
+     */
+    public void cleanup(RoutingNodes routingNodes) {
+        undesiredAllocations.removeAll(shardRouting -> {
+            final var allocationId = shardRouting.allocationId();
+            if (allocationId != null) {
+                return routingNodes.getByAllocationId(shardRouting.shardId(), allocationId.getId()) == null;
+            } else {
+                assert false : "Unassigned shards shouldn't be marked as undesired";
+                return true;
+            }
+        });
+    }
+
+    /**
+     * Clear all tracked allocations
+     */
+    public void clear() {
+        undesiredAllocations.clear();
+    }
+
+    /**
+     * If there are shards that have been in undesired allocations for longer than the configured
+     * threshold, log a warning
+     */
+    public void maybeLogUndesiredShardsWarning(
+        RoutingNodes routingNodes,
+        RoutingAllocation routingAllocation,
+        DesiredBalance desiredBalance
+    ) {
+        final long currentTimeMillis = timeProvider.relativeTimeInMillis();
+        long earliestUndesiredTimestamp = Long.MAX_VALUE;
+        for (var allocation : undesiredAllocations) {
+            if (allocation.value < earliestUndesiredTimestamp) {
+                earliestUndesiredTimestamp = allocation.value;
+            }
+        }
```

Comment:
Can we track this as part of add/remove tracking shards?

Reply:
I'm not clear on whether that would be cheaper. Any time we removed the earliest entry we'd need to do a scan like above to find the next earliest. That could end up happening multiple times when doing a reconciliation. Doing it here means it happens at most once per reconciliation.